unitary/examples/tictactoe/tic_tac_toe.py

```diff
@@ -227,9 +229,113 @@ def sample(self, count: int = 1) -> List[str]:
+            for testing.
+
+    """
+    def __init__(self, game: TicTacToe, file: io.IOBase = sys.stdout):
+        self.game = game
+        self.file = file
+        self.player = "X"
+        self.player_quit = False
+
+    def get_move(self) -> str: 
+        """
+        Gets and returns the player's move.
+
+        Basically a wrapper around input to facilitate testing.
+        """
+        return input(f"Player {self.player} to move (\"help\" for help): ")
+    
+    def player_move(self) -> None:
+        """
+        Interprets the player's move and takes the appropriate action.
+
+        A move can be a one or two letter string within the set [abcdefghi],
+        in which case this function hands the move off to the TicTacToe instance,
+        or one of the GameMoves enums (GameMoves.MAP, GameMoves.EXIT, GameMoves.HELP),
+        which prevent the game loop from alternating to the next player.
+
+        Raises:
+            ValueError if it is still the player's move.
+        """
+        move = self.get_move()
+        print(move)
+
+        if move == GameMoves.MAP.name.lower():
+            print(self.print_board_map(), file=self.file)
+            print("Still your move.", file=self.file)
+            return
+        if move == GameMoves.EXIT.name.lower():
+            self.player_quit = True
+            print("Goodbye!", file=self.file)
+            return
+        if move == GameMoves.HELP.name.lower():
+            print(self.print_help(), file=self.file)
+            print("Still your move.", file=self.file)
+            return
+
+        mark = TicTacSquare.X if self.player == "X" else TicTacSquare.O
+        self.game.move(move, mark)
+        print(self.print_board(), file=self.file)
+        self.player = "O" if self.player == "X" else "X"
+
+    def print_welcome(self) -> str:
+        """
+        Prints the welcome message for the game interface.
+        """
+        message = """
+        Welcome to quantum tic tac toe!
+        Here is the board:
+        """
+        message += self.print_board_map()
+        return message
+
+    def play(self) -> None:
+        """
+        Run the game loop, requesting player moves, alternating players, until
+        the TicTacToe instance reports that the game ends with a winner or a tie
+        or one of the players has quit.
+        """
+        print(self.print_welcome(), file=self.file)
+        while self.game.result() == TicTacResult.UNFINISHED and not self.player_quit: 
+            self.player_move()
+        print(self.game.result(), file=self.file)
+    
+    def print_help(self) -> str:
+        """Print the available moves that player can take."""
+        return """
+    You can enter:
+    - 1 character from [abcdefghi] to place a mark in the corresponding square (eg "a")
+    - 2 characters from [abcdefghi] to place a split mark in corresponding squares (eg "bd")
+    - "map": show board map
+    - "exit" to quit
+    """
```

Comment:
Maybe this should be a constant instead of a function.

Reply:
True. Hold over from when these actually did the printing.